src/main/java/codeu/model/store/basic/MessageStore.java

```diff
@@ -56,37 +59,115 @@ public static MessageStore getTestInstance(PersistentStorageAgent persistentStor
     this.persistentStorageAgent = persistentStorageAgent;
-    messages = new ArrayList<>();
+    messages = new HashMap<>();
+
+    conversationIdToMessageIdList = new HashMap<>();
+    authorIdToMessageIdList = new HashMap<>();
   }
 
-  /** Add a new message to the current set of messages known to the application. */
+  /**
+   * Add a new message to the current set of messages known to the application.
+   * This writes the message to the persistent storage.
+   */
   public void addMessage(Message message) {
-    messages.add(message);
+    addMessageWithoutPersistentStorage(message);
+
     persistentStorageAgent.writeThrough(message);
   }
 
+  /**
+   * Add a new message to the current set of messages known to the application.
+   * This does NOT write the message to the persistent storage.
+   */
+  public void addMessageWithoutPersistentStorage(Message message) {
+    UUID id = message.getId();
+    UUID conversationId = message.getConversationId();
+    UUID authorId = message.getAuthorId();
+
+    messages.put(id, message);
+
+    // Add to its conversation's list
+    if (conversationIdToMessageIdList.containsKey(conversationId)) {
+      conversationIdToMessageIdList.get(conversationId).add(id);
+    } else {
+      conversationIdToMessageIdList.put(conversationId, new ArrayList<>());
+      conversationIdToMessageIdList.get(conversationId).add(id);
+    }
+
+    // Add to its author's list
+    if (authorIdToMessageIdList.containsKey(authorId)) {
+      conversationIdToMessageIdList.get(authorId).add(id);
+    } else {
+      conversationIdToMessageIdList.put(authorId, new ArrayList<>());
+      conversationIdToMessageIdList.get(authorId).add(id);
+    }
+  }
+
   /** Access the current set of Messages within the given Conversation. */
   public List<Message> getMessagesInConversation(UUID conversationId) {
 
-    List<Message> messagesInConversation = new ArrayList<>();
+    List<Message> messagesInConversation;
 
-    for (Message message : messages) {
-      if (message.getConversationId().equals(conversationId)) {
-        messagesInConversation.add(message);
-      }
+    if (conversationIdToMessageIdList.containsKey(conversationId)) {
+      messagesInConversation = // To get the messages in a conversation,
+
+          // We get a list of the message IDs
+          conversationIdToMessageIdList.get(conversationId)
+              // We turn the list into a stream
+              .stream()
+
+              // Then we turn each ID into a message
+              .map(id -> messages.get(id))
+
+              // Then we put the results in a list
+              .collect(Collectors.toList());
+    } else {
+      // If the conversation has no messages, return an empty list
+      messagesInConversation = new ArrayList<>();
     }
 
     return messagesInConversation;
   }
 
+  /** Access the set of Messages sent by a given User. */
+  public List<Message> getMessagesByUser(UUID authorId) {
+
```

Comment:
nit: remove whitespace from beginning of block

Reply:
Done.